libraries/AP_Math/scurves.cpp

```diff
@@ -0,0 +1,645 @@
+#define ALLOW_DOUBLE_MATH_FUNCTIONS
+
+/*
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_Math/AP_Math.h>
+#include "scurves.h"
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL &hal;
+
+// constructor
+scurves::scurves()
+{
+    segtype = SegmentType::EMPTY;
+    otj = 0.0;
+    jerk_max = 0.0;
+    accel_max = 0.0;
+    vel_max = 0.0;
+    _t = 0.0;
+    num_segs = 0;
+}
+
+scurves::scurves(float tj, float Jp, float Ap, float Vp) :
+    otj(tj), jerk_max(Jp), accel_max(Ap), vel_max(Vp)
+{
+    _t = 0.0;
+    num_segs = 0;
+    segtype = SegmentType::EMPTY;
+}
+
+// initialise the S-curve track
+bool scurves::init()
+{
+    segtype = SegmentType::EMPTY;
+    _t = 0.0f;
+    num_segs = 0;
+    add_segment(0.0f, jtype_t::CONSTANT, 0.0f, 0.0f, 0.0f, 0.0f);
+    _track.zero();
+    _delta_unit_1.zero();
+    _delta_unit_2.zero();
+    _delta_unit_3.zero();
+
+    return is_positive(otj) && is_positive(jerk_max) && is_positive(accel_max) && is_positive(vel_max);
+}
+
+// generate an optimal jerk limited curve in 3D space that moves over a straight line between two points
+void scurves::calculate_straight_leg(const Vector3f &origin, const Vector3f &destination)
+{
+    if (!init()) {
+        // Some parameters have been set to zero
+        return;
+    }
+
+    _track = destination - origin;
+    float track_length = _track.length();
+    if (is_zero(track_length)) {
+        // avoid possible divide by zero
+        _delta_unit_1.zero();
+    } else {
+        _delta_unit_1 = _track.normalized();
+        segtype = SegmentType::STRAIGHT;
+        add_segments_straight(track_length);
+    }
+}
+
+// generate jerk limited curve in 3D space approximating a spline between two points
+void scurves::calculate_spline_leg(const Vector3f &origin, const Vector3f &destination, Vector3f origin_vector, Vector3f destination_vector)
+{
+    if (!init()) {
+        // Some parameters have been set to zero
+        return;
+    }
+
+    _track = destination - origin;
+    if (_track.is_zero()) {
+        return;
+    } else if (origin_vector.is_zero() && destination_vector.is_zero()) {
+        calculate_straight_leg(origin, destination);
+        return;
+    }
+
+    float track_length = _track.length();
+    // x is the length of the origin and destination vector assuming a unit length track
```

Comment:
In case it helps, the origin is the beginning of the segment and the destination is the end.  the can be offsets from the EKF origin but it doesn't really matter because the scurve library internally calculates all the positions, velocities and accelerations as relative to the origin